### PR TITLE
Migrate to modern TBB version

### PR DESCRIPTION
**Pull request by SiarheiFedartsou**
# Issue

closes https://github.com/Project-OSRM/osrm-backend/issues/6297
closes https://github.com/Project-OSRM/osrm-backend/issues/6181
closes https://github.com/Project-OSRM/osrm-backend/issues/6087

## Tasklist

 - [ ] CHANGELOG.md entry ([How to write a changelog entry](http://keepachangelog.com/en/1.0.0/#how))
 - [ ] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki)
 - [ ] add tests (see [testing documentation](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/testing.md))
 - [ ] review
 - [ ] adjust for comments
 - [ ] cherry pick to release branch

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: .github/workflows/osrm-backend.yml [hunk 1]
```diff
@@ -155,7 +155,8 @@ jobs:
             CXXCOMPILER: g++-9
             CXXFLAGS: "-m32 -msse2 -mfpmath=sse"
             TARGET_ARCH: i686
-
+            ENABLE_CONAN: ON
```
Comment (SiarheiFedartsou) [inline .github/workflows/osrm-backend.yml:158]: 32-bit build was complaining about some linker errors related to TBB(even though there is 32-bit binary and I also tried to build it from source), so it is easier to simply use Conan instead which seems to work.

File: CMakeLists.txt [hunk 2]
```diff
@@ -459,22 +458,42 @@ if(ENABLE_CONAN)
...
+      bzip2/${CONAN_BZIP2_VERSION}
+      expat/${CONAN_EXPAT_VERSION}
+      lua/${CONAN_LUA_VERSION}
+      onetbb/${CONAN_TBB_VERSION}
+      BASIC_SETUP 
+      BUILD missing 
+      GENERATORS cmake_find_package
+      KEEP_RPATHS
+      NO_OUTPUT_DIRS
+      ARCH x86
```
Comment (SiarheiFedartsou) [inline CMakeLists.txt:480]: I am not CMake pro, so not sure how to reduce copy-paste here(the difference between branches is only this `ARCH x86` line)

File: cmake/FindTBB.cmake [hunk 3]
```diff
@@ -1,286 +1,456 @@
-# Locate Intel Threading Building Blocks include paths and libraries
```
Comment (SiarheiFedartsou) [inline cmake/FindTBB.cmake:1]: Took it from here https://github.com/Kitware/VTK/blob/master/CMake/FindTBB.cmake

File: src/contractor/contractor.cpp [hunk 4]
```diff
@@ -35,12 +35,7 @@
 #include <vector>
 
 #include <boost/assert.hpp>
-
-#if TBB_VERSION_MAJOR == 2020
```
Comment (SiarheiFedartsou) [inline src/contractor/contractor.cpp:39]: https://oneapi-src.github.io/oneTBB/main/tbb_userguide/Migration_Guide.html

File: include/util/packed_vector.hpp [hunk 5]
```diff
@@ -89,6 +92,17 @@ inline WordT set_upper_value(WordT word, WordT mask, std::uint8_t offset, T valu
     return (word & ~mask) | ((static_cast<WordT>(value) >> offset) & mask);
 }
 
+inline bool compare_and_swap(uint64_t *ptr, uint64_t old_value, uint64_t new_value)
```
Comment (SiarheiFedartsou) [inline include/util/packed_vector.hpp:95]: Tbh I am not sure if we need PackedVector to be thread-safe at all - after brief look I couldn't find a place where we modify it concurrently.

File: CMakeLists.txt [hunk 6]
```diff
@@ -496,7 +516,7 @@ if(ENABLE_CONAN)
   set(Boost_UNIT_TEST_FRAMEWORK_LIBRARY "${Boost_unit_test_framework_LIB_TARGETS}")
   
   find_package(BZip2 REQUIRED EXACT ${CONAN_BZIP2_VERSION})
-  find_package(EXPAT REQUIRED EXACT ${CONAN_EXPAT_VERSION})
+  find_package(EXPAT REQUIRED)
```
Comment (SiarheiFedartsou) [inline CMakeLists.txt:519]: https://github.com/Project-OSRM/osrm-backend/pull/6299

**Comment by SiarheiFedartsou** [comment]
@mjjbell hi, would be great to have your opinion here

**Comment by SiarheiFedartsou** [comment]
I see that Windows build fails. Most likely because it uses old TBB [stored on wolt.com](https://github.com/Project-OSRM/osrm-backend/blob/e7185b4bcb7eb17ea4f85dc3d087880dbd309646/appveyor-build.bat#L55). What I am thinking about is may be migrate it to use Conan too - should be probably a bit better approach rather than try to find someone from Wolt to update it now... Another option is to simply install TBB to the system just like I did for Linux/Mac.

**Comment by SiarheiFedartsou** [comment]
> I see that Windows build fails. Most likely because it uses old TBB [stored on wolt.com](https://github.com/Project-OSRM/osrm-backend/blob/e7185b4bcb7eb17ea4f85dc3d087880dbd309646/appveyor-build.bat#L55). What I am thinking about is may be migrate it to use Conan too - should be probably a bit better approach rather than try to find someone from Wolt to update it now... Another option is to simply install TBB to the system just like I did for Linux/Mac.

Btw worth considering migration to GitHub actions for Windows CI 🤔

**Comment by mjjbell** [comment]
Agreed, adding a Windows build to Github Actions CI and killing off the old one is probably the direction to take and will make it easier to perform library upgrades.
IIRC the motivation for moving to C++17 was to take advantage of the STL parallel algorithms. Removing TBB completely could solve these problems too, although might require a lot of surgery.

**Comment by SiarheiFedartsou** [comment]
> IIRC the motivation for moving to C++17 was to take advantage of the STL parallel algorithms. Removing TBB completely could solve these problems too, although might require a lot of surgery.

Yep, but it turned out that there are a lot of compilers that still do not support STL parallel algorithms from C++17 in 2022. One example is Apple Clang. I haven't dug deep here though, but lets first migrate TBB in order to unblock local builds without Conan - it seems to be easier problem anyway.

File: src/storage/storage.cpp [hunk 7]
```diff
@@ -290,27 +290,27 @@ int Storage::Run(int max_wait, const std::string &dataset_name, bool only_metric
 
 std::vector<std::pair<bool, boost::filesystem::path>> Storage::GetStaticFiles()
 {
-    constexpr bool REQUIRED = true;
-    constexpr bool OPTIONAL = false;
+    constexpr bool IS_REQUIRED = true;
+    constexpr bool IS_OPTIONAL = false;
```
Comment (SiarheiFedartsou) [inline src/storage/storage.cpp:294]: `windows.h` defines `OPTIONAL` macro 🤦🏻

File: docker/Dockerfile [hunk 8]
```diff
@@ -4,13 +4,23 @@ ARG BUILD_CONCURRENCY
 RUN mkdir -p /src  && mkdir -p /opt
 
 RUN apt-get update && \
-    apt-get -y --no-install-recommends install cmake make git gcc g++ libbz2-dev libxml2-dev \
-    libzip-dev libboost1.74-all-dev lua5.2 liblua5.2-dev libtbb-dev -o APT::Install-Suggests=0 -o APT::Install-Recommends=0
+    apt-get -y --no-install-recommends install ca-certificates cmake make git gcc g++ libbz2-dev libxml2-dev wget \
+    libzip-dev libboost1.74-all-dev lua5.2 liblua5.2-dev -o APT::Install-Suggests=0 -o APT::Install-Recommends=0
+
+RUN NPROC=${BUILD_CONCURRENCY:-$(nproc)} && \
```
Comment (SiarheiFedartsou) [inline docker/Dockerfile:10]: We could download pre-built binaries, but TBB does not provide arm64 binaries at the moment

**Comment by SiarheiFedartsou** [comment]
@mjjbell I think I am done with it

File: include/util/packed_vector.hpp (hunk 5 above)
Comment (mjjbell) [reply include/util/packed_vector.hpp:95]: According to #4199 it can happen during edge updates, probably when running `osrm-customize`

**Review by mjjbell — APPROVED** [review]
LGTM. Thanks for working though all the required changes :+1:

File: include/util/packed_vector.hpp (hunk 5 above)
Comment (SiarheiFedartsou) [reply include/util/packed_vector.hpp:95]: As I can see set_value is used on push_back to - curious if we are safe on reallocations there 🤔

**Pull request merged**